from os.path import realpath
from os.path import split as pathsplit
from os.path import basename
import subprocess
from snakemake.remote.FTP import RemoteProvider as FTPRemoteProvider
from snakemake.remote.HTTP import RemoteProvider as HTTPRemoteProvider
import sys

# Block annoying warnings
if not sys.warnoptions:
    import warnings
    warnings.simplefilter("ignore")

# Meta
__author__ = "Matt Lawlor"

# Setup
shell.executable("/bin/bash")


# Remotes
HTTP = HTTPRemoteProvider()
FTP = FTPRemoteProvider()

# DETERMINE REMOTE OR LOCAL RESOURCE
def determine_resource(path):
    if "ftp://" in path:
         return FTP.remote(path, immediate_close = True)
    elif "http://" in path:
         return HTTP.remote(path)
    elif "https://" in path:
         return HTTP.remote(path)
    else:
        return path

singularity: "docker://continuumio/miniconda3"
configfile: "config/config.yaml"
pepfile: "config/pep.yaml"

rule all:
    input:
        # The first rule should define the default target files
        # Subsequent target rules can be specified below. They should start with all_*.


rule all_custom_genome:
    input:
        "results/custom-genome/combined.fixed.gtf",
        "results/custom-genome/combined.fasta"

rule all_cellranger:
    input:
        expand("results/cellranger/{s}.done", s=[x.sample_name for x in pep.samples if x.assay == "10X"]),
        #expand("results/fastq-dump-10x/{s}/", s=[x.sample_name for x in pep.samples if x.assay == "10X"])

rule all_scanpy:
    input:
        expand("results/scanpy/{g}/clusters-to-celltypes.csv",g = config.get('SCRNA_GROUPS'))
        #"results/umap.csv.gz",
        #expand("results/ica/{k}/consensus-ica-modules.json",k=COMPS_ICA),
        #expand("results/ica/{k}/consensus-ica-enrichment-{ont}.csv.gz",k=COMPS_ICA,ont=ONTS),
        #expand("results/ica/{k}/consensus-ica-modules-te-types.tsv",k=COMPS_ICA)


rule all_isoseq_1:
    input:
        #expand("results/fastq-dump/{s}/",s=[x.sample_name for x in pep.samples if x.assay == "ISOseq"]),
        #expand("results/isoseq-1/mm2/{s}.{z}", z=['sam'],s=[x.sample_name for x in pep.samples if x.assay == "ISOseq"]),
        #expand("results/isoseq-1/flnc/{s}.bam",s=[x.sample_name for x in pep.samples if x.assay == "ISOseq"]),
        #expand("results/fastqc/{s}",s=[x.sample_name for x in pep.samples if x.assay == "ISOseq"])
        #expand("results/isoseq-1/pbmm2/{s}/{s}.bam",s=[x.sample_name for x in pep.samples if x.assay == "ISOseq"]),
        expand("results/isoseq-1/isoseq_collapse/{s}.gff",s=[x.sample_name for x in pep.samples if x.assay == "ISOseq"]),
        #expand("results/isoseq-1/cupcake-fusion/{s}/{s}.fusion.gff",s=[x.sample_name for x in pep.samples if x.assay == "ISOseq"]),

rule all_chipseq:
    input:
        expand("results/chipseq/aln/{s}/{s}.rmdup.bam.bai",s=[x.sample_name for x in pep.samples if x.assay == "CHIPseq"]),
        #expand("results/chipseq/aln/{s}/{s}_input.bam",s=[x.sample_name for x in pep.samples if x.assay == "CHIPseq"]),

rule all_srnaseq:
    input:
        #expand("results/srna/trimmed/{s}/{s}.srt.bam.bai",s=[x.sample_name for x in pep.samples if x.assay == "sRNAseq"]),
        #expand("results/fastqc/{s}",s=[x.sample_name for x in pep.samples if x.assay == "sRNAseq"]),
        expand("results/small-rna/shortstack-clean/{s}/{s}.bam.bai",s=[x.sample_name for x in pep.samples if x.assay == "sRNAseq"]),


include: "rules/custom-genome.smk"
include: "rules/cellranger.smk"
include: "rules/common.smk"
include: "rules/isoseq-1.smk"
include: "rules/chipseq.smk"
include: "rules/small-rnas.smk"
include: "rules/process-with-scanpy.smk"
