
# Section 3

```{r f3a, fig.cap='Figure 3A: A TE-enriched GEP is highly expressed in a spermatocyte population.'}
w1118.gep_usage %>%
  filter(cons %in% tep.name) %>%
  collect() %>%
  left_join(collect(w1118.obs)) %>%
  arrange(usage) %>%
  ggplot(aes(X_umap1,X_umap2, color=usage)) +
  geom_point(size=0.75) +
  scale_color_gradient2(low='steelblue', mid='lightgray',high='tomato', midpoint = 0.04) +
  theme_classic() +
  coord_fixed()
```


```{r f3b, fig.cap='Figure 3B: Shown are all GEPs with  >1 TE. The TEP is highly enriched for TEs'}
w1118.gep_membership %>%
  filter(qval < 0.005) %>%
  collect() %>%
  left_join(te.lookup, by=c(X1='merged_te')) %>%
  filter(!str_detect(X1,'FBgn')) %>%
  dplyr::select(module, X1, repClass, repFamily) %>%
  distinct() %>%
  group_by(module) %>%
  mutate(n_tes=n()) %>%
  filter(n_tes > 1) %>%
  mutate(repClass=replace_na(repClass,replace = 'other')) %>%
  mutate(module=paste0("GEP-",module)) %>%
  ggplot(aes(reorder(module, n_tes), fill=repClass)) +
  geom_bar(position='stack') +
  scale_fill_brewer(type='qual',palette=6, name='') +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90, hjust = 1, vjust=0.5)) +
  #theme(aspect.ratio = 0.3,legend.position = 'bottom', axis.text.x = element_text(angle=45, hjust=90, vjust=0.5)) +
  xlab("") + ylab('N') +
  scale_y_continuous(expand=c(0,0))
```

```{r f3c, fig.cap='Figure 3C'}
feature_types <- c('mRNA','ncRNA','snRNA','pre_miRNA','miRNA','pseudogene','snoRNA')
feature_df <- gtf %>% 
  mutate(type=as.character(type)) %>%
  filter(type %in% feature_types) %>%
  dplyr::select(gene_id,type) %>%
  distinct()

w1118.gep_membership %>%
  filter(qval < 0.005) %>%
  collect() %>%
  left_join(feature_df, by=c('X1'='gene_id')) %>%
  filter(module==tep.name) %>%
  mutate(type=ifelse(!str_detect(X1,'FBgn'),'TE',type)) %>%
  count(type) %>%
  mutate(pct=n/sum(n)) %>%
  ggplot(aes("",pct,fill=type)) +
  geom_bar(color='white',stat='identity') +
  coord_polar('y',start=0) +
  scale_fill_brewer(type='qual',palette=6,direction = -1, name='') +
  theme_no_axes() +
  theme(aspect.ratio = 1, legend.position='bottom') +
  xlab('') + ylab('') +
  geom_text(aes(label = scales::percent(round(pct,3))), position = position_stack(vjust = 0.5))
```

```{r}
w1118.usage.mat <- w1118.gep_usage %>% 
  collect() %>%
  #group_by(X1) %>%
  #mutate(usage = scales::rescale(usage,c(-1,1))) %>%
  spread(cons, usage) %>%
  column_to_rownames('X1') %>% as.matrix()

#sdum <- sd(w1118.usage.mat)
#w1118.usage.mat <- t(apply(w1118.usage.mat, MARGIN = 1, FUN=function(x) (x-mean(x))/sdum))

w1118.usage.hc.geps <- w1118.usage.mat %>% cor() %>% {1-.} %>% as.dist() %>% hclust()
w1118.usage.hc.cells <- w1118.usage.mat %>% t() %>% cor() %>% {1-.} %>% as.dist() %>% hclust()
```

```{r}
w1118.membership.mat <- w1118.gep_membership %>% 
  collect() %>%
  dplyr::select(-qval) %>%
  spread(module, weight) %>%
  column_to_rownames('X1') %>% as.matrix()

w1118.membership.hc.geps <- w1118.membership.mat  %>% cor() %>% {1-.} %>% as.dist() %>% hclust()
w1118.membership.hc.genes <- w1118.membership.mat  %>% t() %>% cor() %>% {1-.} %>% as.dist() %>% hclust()
```

```{r s3a, out.height="4in", fig.dim=c(10,10), fig.cap='Figure S3A'}
palette_len <- 255
colors <- colorRampPalette( (c("#3B9AB2", "#3B9AB2","#78B7C5", "#E1AF00", "#F21A00")))(palette_len)
  
myBreaks <- c(seq(min(w1118.usage.mat)*0.1, mean(w1118.usage.mat), length.out=ceiling(palette_len/2)),
              seq(median(w1118.usage.mat) + 0.01, max(w1118.usage.mat)*0.1, length.out=floor(palette_len/2)))

cls <- circlize::colorRamp2(breaks = myBreaks,colors = colors)

levs <- w1118.obs %>% collect() %>%
  pull(clusters2) %>% unique %>%
  set_names(.,.) %>%
  map(~str_remove(.,'\\d+\\/')) %>%
  {names(.)[order(unlist(.))]}

col <- list(cell_types=RColorBrewer::brewer.pal(length(levs),'Set3') %>%
               set_names(levs))

  # Create the heatmap annotation
ha <- HeatmapAnnotation(cell_types = pull(w1118.obs,'clusters2'), col = col)
  
Heatmap(t(w1118.usage.mat)[,pull(w1118.obs,'X1')],
                 column_names_rot = 45,
                 column_title_side = "bottom",
                 na_col = 'green',
                 cluster_rows = w1118.membership.hc.geps,
                 name = 'GEP usage',
                  row_title = 'GEPs', 
                 #top_annotation = ha,
                 show_column_names = F,
                 col = cls,
                  column_title_rot = 90,
                 column_title_gp = gpar(fontsize=12),
                 row_names_gp = gpar(fontsize = 2),
                  show_row_names = T,
                 show_column_dend = F, 
                show_row_dend = T,
                 column_split = factor(pull(w1118.obs,'clusters2'),levels=levs),
                 use_raster = T,
                 heatmap_height = unit(4.7,units = "in"),
                 width=unit(8.7,units="in"),
                 cluster_column_slices = F)
```

```{r s3b, fig.cap='Supp. Fig. 3B'}
metrics.grid <- w1118.grid %>%
  collect() %>%
  mutate_at(vars('cov','pct_unique_term','pct_enr'), scales::rescale) %>%
  mutate(score = cov * pct_unique_term)

metrics.grid %>%
 dplyr::rename(coverage='cov', uniqueness='pct_unique_term') %>%
 gather(metric, value, -rep:-comps) %>%
 filter(metric %in% c("coverage","uniqueness")) %>%
ggplot(aes(comps, value, color=metric))  +
geom_smooth() +
geom_point() +
facet_grid(qval~rep) +
theme_classic() +
scale_color_viridis_d(direction=-1) +
theme(aspect.ratio = 0.5, plot.caption = element_text(hjust=0)) +
xlab("k") + ylab('score')
```


```{r s3c, fig.cap='Supp. Fig. 3C'}
ggplot(metrics.grid,aes(as.factor(comps),as.factor(qval),fill=score)) +
  geom_raster(interpolate = F) +
  theme_minimal() +
  xlab("k") + ylab("qval") +
  scale_fill_viridis_c() +
  theme(aspect.ratio = 1, plot.caption = element_text(hjust=0)) +
  labs(caption = "Joint score displayed by FDR cutoff and k.") +
  facet_wrap(~rep)
```

```{r s3d, fig.cap='Supp. Fig. 3D'}
w1118.gep_membership %>%
  filter(qval < optimal_qval) %>%
  collect() %>%
  group_by(module) %>% tally() %>%
  ggplot(aes(n)) +
  geom_histogram(aes(fill=..count..), color='black') +
  theme_classic() +
  scale_y_continuous(expand = c(0,0)) +
  scale_fill_fermenter(direction = 1, palette = 13) +
  theme(aspect.ratio = 0.5, plot.caption = element_text(hjust=0)) +
  xlab('module sizes')
```


```{r eval=F}
grid.dups <- Sys.glob(paste0("../results/gep-grid-search/larval-w1118-testes/ica/*/*/consensus-ica.csv.gz")) %>%
  tibble(file=.) %>%
  mutate(k=str_extract(file,"(?<=ica\\/)\\d+(?=\\/)")) %>%
  mutate(rep=str_extract(file,"(?<=\\/)\\d+(?=\\/consensus)")) %>%
  mutate(data = map(file, read_csv)) %>%
  mutate(data = map(data,column_to_rownames,'index')) %>%
  mutate(data = map(data,cor)) %>%
  mutate(data = map(data,as_tibble, rownames='module')) %>%
  mutate(data = map(data,gather,module2, corr, -module)) %>%
  unnest(cols=data) %>%
  filter(module!=module2) %>%
  group_by(k,rep,module) %>%
  summarize(closest = max(corr)) %>%
  group_by(k,rep) %>%
  summarize(dup.rate = sum(closest>=0.9)/n())
```


```{r s3e, fig.cap='Supp. Fig. 3E'}
usage <- w1118.gep_usage %>%
  collect() %>%
  spread(cons, usage) %>%
  column_to_rownames('X1')

usage.cor <- cor(usage, method='spearman')

usage.cor.hc <- {1-usage.cor} %>% as.dist() %>% hclust


pheatmap::pheatmap(usage.cor,
                   cluster_rows = usage.cor.hc, 
                   cluster_cols = usage.cor.hc, treeheight_row = F,treeheight_col = 10,fontsize_col = 2,
                   border_color = NA, cellwidth = 2, cellheight = 2, show_rownames = F)
```

```{r s3f, fig.cap='Supp. Fig. 3F', dev.args = list(ragg_png= list(scaling=1))}
rnaseq_pe_df <- collect(rnaseq_pe) %>% filter(strandedness == "first_strand")

hsbam_dat <- left_join(rnaseq_pe_df, te.lookup) %>%
  mutate(gene_id = ifelse(!is.na(merged_te),merged_te,gene_id)) %>%
  group_by(sample,gene_id) %>%
  summarize(count = sum(count)) %>%
    spread(sample,count) %>%
  column_to_rownames('gene_id') %>%
  as.matrix()

hsbam_cd <- tibble(name=colnames(hsbam_dat )) %>%
mutate(condition = str_remove(str_remove(name,c('RNA_seq_')),'_rep.+')) %>%
  column_to_rownames('name')

gtf %>%
  filter(type == 'exon') %>%
  left_join(dplyr::select(te.lookup,gene_id,merged_te)) %>%
  mutate(gene_id= ifelse(!str_detect(gene_id,'FBgn'),merged_te,gene_id)) %>% # grouping by merged_te
  GRanges() %>%
  split(.,.$gene_id) -> grx

stopifnot(length(rownames(hsbam_dat)[!rownames(hsbam_dat) %in% names(grx)])==0)

hsbam_dds <- grx %>%
  .[rownames(hsbam_dat)] %>%
  SummarizedExperiment(assays = list(counts=hsbam_dat), rowRanges=., colData=hsbam_cd) %>%
  DESeqDataSet(design = ~condition)

fpkm_df <- fpkm(hsbam_dds) %>%
  as_tibble(rownames = 'gene') %>%
  gather(condition, fpkm, -gene) %>%
  mutate(condition = str_remove(condition,'_rep\\d+')) %>%
  mutate(GEP = as.factor(ifelse(gene %in% tep.members,'TEP','other'))) %>%
  mutate(type = as.factor(ifelse(str_detect(gene,'FBgn'),'gene','TE'))) %>%
  group_by(gene, condition, GEP, type) %>%
  summarize(fpkm=mean(fpkm)) %>%
  ungroup() %>%
  mutate(condition = str_remove(condition,'RNA_seq_')) %>%
  mutate(condition = str_remove(condition,'_mutant')) %>%
   mutate(condition = fct_relevel(condition,  c('bam','aly',
                    '48hrPHS','72hrPHS')))

fpkm_df %>%
ggplot(aes(condition, fpkm, fill=condition)) +
  geom_boxplot(position=position_dodge(), outlier.shape = NA) +
  #geom_violin(position=position_dodge(), scale = 'width') +
  theme_classic() +
  theme(aspect.ratio = 0.4, axis.text.x = element_text(angle=90, hjust=1, vjust=0.5), legend.position = 'bottom') +
  coord_cartesian(ylim = quantile(fpkm_df$fpkm, c(0.001, 0.95))) +
  scale_fill_brewer(type='qual',palette = 6) +
  facet_grid(type~GEP, scales = 'free_y') +
  ggpubr::stat_compare_means(aes(label = paste0("p = ", ..p.format..)), label.y = 125) +
  guides(fill=F) +
  xlab("")
  
```


