---
title: "Figures"
author: ""
date: "4/27/2021"
output: 
  pdf_document:
    latex_engine: xelatex
sansfont: Arial
header-includes:
- \AtBeginDocument{\let\maketitle\relax}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, warning = F, message = F)
```

```{r, include=F}
library(tidyverse)
library(patchwork)
library(ggtext)
library(magick)

extrafont::loadfonts(quiet=TRUE)
```

\newpage

```{r, fig.dim=c(10,6), out.height="4.2in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/figure1.rds")
```

*Figure 1: Identification of testis cell populations.* **A)** UMAP projection groups transcriptionally similar cells in 2D space. Cells are colored by assigned cell type. **B)** Dot plot shows expression of selected marker genes used for cell type assignment. Color of each dot corresponds to mean normalized and log-transformed expression within cell clusters. Dot size corresponds to the proportion of cells in each cluster expressing the marker. **C)** Cell counts within each cell type cluster. Spermatocytes are the most abundant cell type by our analysis approach. 

\newpage

```{r}
read_rds("~/work/TestisTEs2021/results/panels/figure2.rds")
```

*Figure 2: A spermatocyte cluster expresses transposons at high levels.* Heatmap shows scaled expression level of all transposable elements detected in this dataset across all cells. Several clusters express small groups of transposable elements uniformly or show sporadic expression of transposons in some member cells. Cluster 3/Spermatocyte shows uniformly high expression of many transposable elements. 

\newpage

```{r, fig.dim=c(10,6), out.height="4.5in", out.width="7.5in"}
read_rds("~/work/TestisTEs2021/results/panels/figure3.rds")
```

*Figure 3: A TE-enriched gene expression program.* **A)** UMAP projection colored by GEP-27 expression score. Expression score is derived from the consensus (averaged) ICA source matrix. **B)** Tallies of TE classes found in each GEP containing at least 1 TE. GEP-27 contains over 300 total features including 75 transposable elements. In comparison, the next most transposable element-enriched GEP contains approximately 20 TEs. GEP-27 TEs are predominantly retrotransposons. **C)** Composition of GEP-27.  GEP-27 is predominantly host-genes. 

\newpage

```{r, fig.dim=c(10,10), out.height="7in", out.width="7in"}

read_rds("~/work/TestisTEs2021/results/panels/figure4.rds")
```

*Figure 4.  TEP co-occurs with Y chromosome transcriptional activity.* **A)** Dot plot shows mean normalized expression of Y-linked genes in each cluster. Dot size corresponds to the proportion of cells in each cluster with detectable expression of each Y-linked gene. Y-linked genes, especially fertility factors kl-3 and kl-5, are highly expressed by cluster 3/Spermatocyte. **B)** Violin plot shows normalized expression of EAChm is largely confined to cluster 3/Spermatocyte. **C)** Normalized expression of ACCORD2 and QUASIMODO2 is largely confined to cluster 3/Spermatocyte. **D)** Multiplexed RNA-FISH in whole-mount 3rd larval instar w1118 testis shows QUASIMODO2 and EAChm expression is detected in the middle region of the testis, where primary spermatocytes are located. Red: EAChm; green: ACCORD2; blue: DAPI. 

\newpage

```{r, fig.dim=c(14,8), out.height="4in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/figure5.rds")
```

*Figure 5. TEP-TEs are enriched on the Y chromosome.* **A)** Boxplot showing per-megabase polymorphic insertions of TEP-TEs versus non-TEP-TEs. TEP-TEs are depleted from the X compared to other TEs. Wilcoxon rank-sum test, P = 0.029. Polymorphic insertion data is derived from TIDAL-Fly DGRP data (see Methods). **B)** A higher proportion of TEP-TE insertions are found on the Y chromosome compared to non-TEP-TEs. Chi-square test, P = 2.29e-292. To better estimate Y-linked insertions despite the incomplete Y-assembly in the reference sequence, insertions were mapped from a heterochromatin-enriched assembly (see Methods). **C)** Ratios of male and female copy number for individual TEs estimated from w1118 WGS coverage. TEP-TEs are present at higher copy number in the male genome Wilcoxon rank-sum test, P=0.0035. This indicates that on average each TEP-TE family is relatively enriched on the Y chromosome, versus non-TEP-TEs. **D)** Ratio of w1118 testis RNA-seq depth to WGS depth for each TEâ€™s most over-expressed Y-linked allele. In comparison, the reference allele is expressed proportionately to DNA copy number (see Supplementary Figures). On average, TEP-TEs have at least one Y-linked allele that is over-expressed relative to DNA copy number, compared to non-TEP-TEs. Wilcoxon rank-sum test, P=1.7e-07. 

\newpage

```{r, fig.dim=c(20,20), out.height="7in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/supp1.rds")
```

*Supplementary Figure 1. scRNA-seq data.* 
**A)** Scatterplots show pseudo-bulk expression derived from our *w1118* scRNA pipeline (see **Methods**) versus bulk expression for 4 *w1118* testis poly-A RNA-seq replicates generated by Mahadevaraju et al. 2021 (see *Methods*). Each replicate shows strong correlation with pseudo-bulk (all Pearson's R >= 0.89, P<2.2e-16). **B)** Scatterplots show same analysis described in **A** but restricted to TEs (all Pearson's R >= 0.85, P<2.2e-16). **C)** Heatmap shows correlation between scRNA-seq expression estimates derived from our pipeline for all clusters identified in this study compared to clusters identified by Mahadevaraju et al. 2021 (source of data). **D)** Barplot shows the number of cells used in Mahadevaraju et al. 2021 (gray) and this study (red). **E)** Barplot shows the number of cells assigned to each cluster. Bars are divided and colored by the scRNA replicate from which cells are derived. **F)** Distribution of doublet scores (see **Methods**) for cells in each cluster after all filtering steps. **G)** Distribution of total log1p(UMIs) for cells in each cluster after all filtering steps. **H)** Barplot shows the number of cells assigned to each cluster. Bars are divided and colored by predicted cell cycle phase (see **Methods**). **I)** Distribution of total detected genes for cells in each cluster after all filtering steps. **J)** Distribution of the percentage of total pre-filtering mitochondrial reads for cells in each cluster after all filtering steps.

\newpage

```{r, fig.dim=c(20,20), out.height="7in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/supp2.rds")
```

*Supplementary Figure 2. TE expression in scRNA.* **A)** Heatmaps show distribution of sense-strand poly-A RNA-seq signal for detected host genes (left) and detected TEs (right) in bins comprising the full length of the respective features. poly-A RNA-seq data generated by Mahadevaraju et al 2021 (see *Methods*). **B)** Boxplots show standard deviations for expression across bins for host genes (gray) and TEs (red). Three of four replicates show no significant difference in variability of poly-A signal across bins within features (Wilcoxon rank-sum test P>0.05). Replicate 3 shows a significant difference (Wilcoxon rank-sum test, P=0.039). **C)** Barplot shows number of genic fusions reproducibly found in poly-A RNA-seq data. **D)** For each TE introduced in **C**, the y-axis position of each point represents the number of uniquely-mapping chimeric reads detected by STAR that support each breakpoint (see **Methods**). **E**) Violin plots show distributions of raw TE-mapping UMI counts in each L3 *w1118* cell cluster. TE counts vary significantly among the clusters (Kruskal-Wallis, p < 2.2e-16). **F**) Violin plots show distributions of depth normalized TE-mapping UMI counts in each L3 *w1118* cell cluster. TE counts vary significantly among the clusters (Kruskal-Wallis, p < 2.2e-16).

\newpage

```{r, fig.dim=c(8,10), out.height="8in", out.width="10in"}
read_rds("~/work/TestisTEs2021/results/panels/supp3.rds")
```

*Supplementary Figure 3. Gene Expression Program detection via consensus approach to Independent Component Analysis.* 
**A)** Heatmaps correspond to independent replicates of the grid search approach used to optimize ICA component number (*k*) and *q*-value cutoff. Color intensity corresponds to the enrichment score (see **Methods**) for each combination of *q* and *k*. **B)** Histogram shows module sizes among the set of GEPs used in main analysis. The majority of modules detected include fewer than 100 features. **C)** Scatterplots show relationship of TE expression in *w1118* Cluster 3/Spermatocyte versus all clusters detected by our pipeline in the "Wild Strain" dataset generated by Witt et al. 2019. Multiple spermatocyte clusters (highlighted in red) show TE expression patterns similar to 3/Spermatocyte.

\newpage

```{r, fig.dim=c(8,8), out.height="7in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/supp4.rds")
```

*Supplementary Figure 4. Redundancy and TE content of GEPs.* **A)** Clustering of GEPs by cell usage score (mixing matrix). **B)** Clustering of GEPs by gene membership score (mixing matrix). For **A**,**BB** `1-Pearson's R` is used as a distance metric. **C)** Breakdown of specific DNA TE, LINE, and LTR TE families in GEP-27.


\newpage

```{r, fig.dim=c(7,8), out.height="4in", out.width="7in", fig.align="center"}
image_read('/media/mlawlor/T7/microscopy_figs/210215_accord2_calfluor610_eachm_quasar670_3p4.slices_1_1.split_channels.png') %>% image_ggplot()
```

*Supplementary Figure 5. ACCORD2 and EAChm expression in L3 testes.*
Representative slice of multiplexed RNA-FISH in whole-mount 3rd larval instar w1118 testis. Image is split by color channel. ACCORD2 and *EAChm* expression is detected in the middle region of the testis, where primary spermatocytes are located. Red: *EAChm*; green: QUASIMODO2; blue: DAPI. 

\newpage

```{r, fig.dim=c(7,8), out.height="5in", out.width="7in", fig.align='center'}
image_read('/media/mlawlor/T7/microscopy_figs/210215_accord2_calfluor610_eachm_quasar670_3p4.slices_1_30.montage.slices_1_30.montage.png') %>% image_ggplot()
```

*Supplementary Figure 6. ACCORD2 and EAChm expression in L3 testes.* 
Z-slices montage of multiplexed RNA-FISH in whole-mount 3rd larval instar w1118 testis. ACCORD2 and *EAChm* expression is detected in the middle region of the testis, where primary spermatocytes are located. Red: *EAChm*; green: QUASIMODO2; blue: DAPI. 

\newpage

```{r, fig.dim=c(7,8), out.height="5in", out.width="7in", fig.align='center'}
image_read('/media/mlawlor/T7/microscopy_figs/210215_quasimodo2_calfluor610_eachm_quasar670_slide2_2p.slices_1_19.montage.png') %>% image_ggplot()
```

*Supplementary Figure 7. QUASIMODO2 and EAChm expression in L3 testes.* Z-slices montage of multiplexed RNA-FISH in whole-mount 3rd larval instar w1118 testis. QUASIMODO2 and *EAChm* expression is detected in the middle region of the testis, where primary spermatocytes are located. Red: *EAChm*; green: QUASIMODO2; blue: DAPI. 

\newpage

```{r, fig.dim=c(10,10), out.height="7in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/supp5.rds")
```

*Supplementary Figure 8. TEP-TEs are enriched on the Y-chromosome.* 
**A)** Dot plot shows expression of selected effectors of spermatocyte transcriptional programs. Color of each dot corresponds to mean normalized and log-transformed expression within cell clusters. Dot size corresponds to the proportion of cells in each cluster expressing the marker. **B)** Bar plot shows proportion of Y chromosome genes or other genes that are assigned to the TE-enriched Program or other GEPs. Chi-square test P=1.7e-05. **C)** Scatter plot shows estimated male copies (x-axis) versus estimated female copies (y-axis) of TEP-TEs and TEs assigned to other GEPs or not assigned to a GEP. **D)** Bars represent the ratios of male to female estimated copies for the top 10 male- and female-enriched TEs. Bars are colored by membership in TEP.
