---
title: "Figures"
author: ""
date: "4/27/2021"
output: 
  pdf_document:
    latex_engine: xelatex
sansfont: Arial
header-includes:
- \AtBeginDocument{\let\maketitle\relax}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, warning = F, message = F)
```

```{r, include=F}
library(tidyverse)
library(patchwork)
library(ggtext)

extrafont::loadfonts(quiet=TRUE)
```

\newpage

```{r, fig.dim=c(10,6), out.height="4.2in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/figure1.rds")
```

*Figure 1: Identification of testis cell populations.* **A)** UMAP projection groups transcriptionally similar cells in 2D space. Cells are colored by assigned cell type. **B)** Dot plot shows expression of selected marker genes used for cell type assignment. Color of each dot corresponds to mean normalized and log-transformed expression within cell clusters. Dot size corresponds to the proportion of cells in each cluster expressing the marker. **C)** Cell counts within each cell type cluster. Spermatocytes are the most abundant cell type by our analysis approach. 

\newpage

```{r}
read_rds("~/work/TestisTEs2021/results/panels/figure2.rds")
```

*Figure 2: A spermatocyte cluster expresses transposons at high levels.* **A)** Heatmap shows scaled expression level of all transposable elements detected in this dataset across all cells. Several clusters express small groups of transposable elements uniformly or show sporadic expression of transposons in some member cells. Cluster 3/Spermatocyte shows uniformly high expression of many transposable elements. 

\newpage

```{r, fig.dim=c(10,6), out.height="4.5in", out.width="7.5in"}
read_rds("~/work/TestisTEs2021/results/panels/figure3.rds")
```

*Figure 3: A TE-enriched gene expression program.* **A)** UMAP projection colored by GEP-27 expression score. Expression score is derived from the consensus (averaged) ICA source matrix. **B)** Tallies of TE classes found in each GEP containing at least 1 TE. GEP-27 contains over 300 total features including 75 transposable elements. In comparison, the next most transposable element-enriched GEP contains approximately 20 TEs. GEP-27 TEs are predominantly retrotransposons. **C)** Composition of GEP-27.  GEP-27 is predominantly host-genes. 

\newpage

```{r, fig.dim=c(10,10), out.height="7in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/figure4.rds")
```

*Figure 4.  TEP co-occurs with Y chromosome transcriptional activity.* **A)** Dot plot shows mean normalized expression of Y-linked genes in each cluster. Dot size corresponds to the proportion of cells in each cluster with detectable expression of each Y-linked gene. Y-linked genes, especially fertility factors kl-3 and kl-5, are highly expressed by cluster 3/Spermatocyte. **B)** Violin plot shows normalized expression of EAChm is largely confined to cluster 3/Spermatocyte. **C)** Normalized expression of ACCORD2 and QUASIMODO2 is largely confined to cluster 3/Spermatocyte. **D)** Multiplexed RNA-FISH in whole-mount 3rd larval instar w1118 testis shows QUASIMODO2 and EAChm expression is detected in the middle region of the testis, where primary spermatocytes are located. Red: EAChm; white: QUASIMODO2. 

\newpage

```{r, fig.dim=c(14,8), out.height="4in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/figure5.rds")
```

*Figure 5. TEP-TEs are enriched on the Y chromosome.* **A)** Boxplot showing per-megabase polymorphic insertions of TEP-TEs versus non-TEP-TEs. TEP-TEs are depleted from the X compared to other TEs. Wilcoxon rank-sum test, P = 0.029. Polymorphic insertion data is derived from TIDAL-Fly DGRP data (see Methods). **B)** A higher proportion of TEP-TE insertions are found on the Y chromosome compared to non-TEP-TEs. Chi-square test, P = 2.29e-292. To better estimate Y-linked insertions despite the incomplete Y-assembly in the reference sequence, insertions were mapped from a heterochromatin-enriched assembly (see Methods). **C)** Ratios of male and female copy number for individual TEs estimated from w1118 WGS coverage. TEP-TEs are present at higher copy number in the male genome Wilcoxon rank-sum test, P=0.0035. This indicates that on average each TEP-TE family is relatively enriched on the Y chromosome, versus non-TEP-TEs. **D)** Ratio of w1118 testis RNA-seq depth to WGS depth for each TEâ€™s most over-expressed Y-linked allele. In comparison, the reference allele is expressed proportionately to DNA copy number (see Supplementary Figures). On average, TEP-TEs have at least one Y-linked allele that is over-expressed relative to DNA copy number, compared to non-TEP-TEs. Wilcoxon rank-sum test, P=1.7e-07. 

\newpage

```{r, fig.dim=c(20,20), out.height="7in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/supp1.rds")
```


\newpage

```{r, fig.dim=c(20,20), out.height="7in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/supp2.rds")
```

\newpage

```{r, fig.dim=c(8,10), out.height="8in", out.width="10in"}
read_rds("~/work/TestisTEs2021/results/panels/supp3.rds")
```

\newpage

```{r, fig.dim=c(8,8), out.height="7in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/supp4.rds")
```

\newpage

```{r, fig.dim=c(10,10), out.height="7in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/supp5.rds")
```
