---
title: "Figures"
author: ""
date: "4/27/2021"
output: 
  pdf_document:
    latex_engine: xelatex
sansfont: Arial
mainfont: Arial
header-includes:
- \AtBeginDocument{\let\maketitle\relax}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, warning = F, message = F, dev = "png", dpi = 200)
```

```{r, include=F}
library(tidyverse)
library(patchwork)
library(ggtext)
library(magick)

extrafont::loadfonts(quiet=TRUE)
```

\newpage

```{r, fig.dim=c(10,6), out.height="4.2in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/figure1.rds")
```

*Figure 1: Identification of testis cell populations.* **A)** UMAP projection groups transcriptionally similar cells in 2D space. Cells are colored by assigned cell type. **B)** Dot plot shows expression of selected marker genes used for cell type assignment. Color of each dot corresponds to mean normalized and log-transformed expression within cell clusters. Dot size corresponds to the proportion of cells in each cluster expressing the marker. **C)** Cell counts within each cell type cluster.

\newpage

```{r}
read_rds("~/work/TestisTEs2021/results/panels/figure2.rds")
```

*Figure 2: A spermatocyte cluster expresses transposons at high levels.* Heatmap shows scaled expression level of all transposable elements detected in this dataset across all cells. Several clusters express small groups of transposable elements uniformly or show sporadic expression of transposons in some member cells. Cluster 3/Spermatocyte shows uniformly high expression of many transposable elements. 

\newpage

```{r, fig.dim=c(10,6), out.height="4.5in", out.width="7.5in"}
read_rds("~/work/TestisTEs2021/results/panels/figure3.rds")
```

*Figure 3: A TE-enriched gene expression program is expressed primarily in cluster 3/Spermatocytes.* **A)** Tallies of TE classes found in each GEP containing at least 1 TE. GEP-27 contains almost four-fold more TEs than the next most TE-rich GEP and is predominately composed of LTR retrotransposons (59%%), LINE (29%) and DNA (9%) elements. **B)** GEP-27 contains over 300 total features, the vast majority of which are either protein-coding genes (68%), TEs (18.6%), or non-coding RNAs (12%). **C)** UMAP projection colored by GEP-27 expression score. Expression score is derived from the consensus (averaged) ICA source matrix. 

\newpage

```{r, fig.dim=c(10,10), out.height="7in", out.width="7in"}

read_rds("~/work/TestisTEs2021/results/panels/figure4.rds")
```

*Figure 4.  TEP co-occurs with Y chromosome transcriptional activity.* **A)** Violin plot shows normalized expression of *EAChm*, *QUASIMODO2*, and *ACCORD2* is largely confined to cluster 3/Spermatocyte.  **B)** Multiplexed RNA-FISH in whole-mount 3rd larval instar w1118 testis shows *ACCORD2* and *EAChm* expression is detected in the middle region of the testis, where primary spermatocytes are located. Red: *EAChm*; green: *ACCORD2*; blue: DAPI. **C)** Dot plot shows mean normalized expression of Y-linked genes in each cluster. Dot size corresponds to the proportion of cells in each cluster with detectable expression of each Y-linked gene. Y-linked genes, especially fertility factors *kl-3* and *kl-5*, are highly expressed by cluster 3/Spermatocyte.   

\newpage

```{r, fig.dim=c(14,8), out.height="4in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/figure5.rds")
```

*Figure 5. TEP-TEs are enriched on the Y chromosome.* **A)** A higher proportion of TEP-TE insertions are found on the Y chromosome compared to non-TEP-TEs. Chi-square test, P = 2.29e-292. To better estimate Y-linked insertions despite the incomplete Y-assembly in the reference sequence, insertions were mapped from a heterochromatin-enriched assembly (see **Methods**). **B)**  Ratios of male and female copy number for individual TEs estimated from w1118 WGS coverage. TEP-TEs are present at higher copy number in the male genome Wilcoxon rank-sum test, P=0.0035. **C)** Allele specific analysis of TE expression (see **Methods**) shows that Y-linked copies of TEP TEs are overexpressed relative to their DNA copy number and this overexpression is significantly larger than that of non-TEP-TEs Wilcoxon rank-sum test, P=1.7e-07.  **D)** Boxplot showing the ratios of X-linked versus autosomal polymorphic insertions for each TE in the TIDAL-fly (ref) database. TEP-TEs are depleted from the X compared to other TEs with polymorphic insertions. Wilcoxon rank-sum test, P = 0.029. **E)**  Dot plot shows expression of selected piRNA pathway genes. Color of each dot corresponds to mean normalized and log-transformed expression within cell clusters. Dot size corresponds to the proportion of cells in each cluster expressing the marker.

\newpage

```{r, fig.dim=c(20,20), out.height="6.5in", out.width="6.5in"}
read_rds("~/work/TestisTEs2021/results/panels/supp1.rds")
```

\small

*Supplementary Figure 1. scRNA-seq data.* 
**A)** Scatterplots show pseudo-bulk expression derived from our *w1118* scRNA pipeline (see **Methods**) versus bulk expression for 4 *w1118* testis poly-A RNA-seq replicates generated by Mahadevaraju et al. 2021 (see **Methods**). Each replicate shows strong correlation with pseudo-bulk (all Pearson's R >= 0.89, P<2.2e-16). **B)** Scatterplots show same analysis described in **A** but restricted to TEs (all Pearson's R >= 0.85, P<2.2e-16). **C)** Heatmap shows correlation between scRNA-seq expression estimates derived from our pipeline for all clusters identified in this study compared to clusters identified by Mahadevaraju et al. 2021 (source of data). **D)** Barplot shows the number of cells used in Mahadevaraju et al. 2021 (gray) and this study (red). **E)** Barplot shows the number of cells assigned to each cluster. Bars are divided and colored by the scRNA replicate from which cells are derived. **F)** Distribution of doublet scores (see **Methods**) for cells in each cluster after all filtering steps. **G)** Distribution of total log1p(UMIs) for cells in each cluster after all filtering steps. **H)** Barplot shows the number of cells assigned to each cluster. Bars are divided and colored by predicted cell cycle phase (see **Methods**). **I)** Distribution of total detected genes for cells in each cluster after all filtering steps. **J)** Distribution of the percentage of total pre-filtering mitochondrial reads for cells in each cluster after all filtering steps.

\newpage

```{r, fig.dim=c(20,20), out.height="7in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/supp2.rds")
```

*Supplementary Figure 2. TE expression in scRNA.* **A**) Violin plots show distributions of raw TE-mapping UMI counts in each L3 *w1118* cell cluster. TE counts vary significantly among the clusters (Kruskal-Wallis, p < 2.2e-16). **B**) Violin plots show distributions of depth normalized TE-mapping UMI counts in each L3 *w1118* cell cluster. TE counts vary significantly among the clusters (Kruskal-Wallis, p < 2.2e-16). **C)** Heatmaps show distribution of sense-strand poly-A RNA-seq signal for single-isoform host gene mRNAs (left) and detected TEs (right) in bins comprising the full length of the respective features. Poly-A RNA-seq data generated by Mahadevaraju et al 2021 (see **Methods**). **D)** Boxplots show standard deviations of expression across bins for host genes (gray) and TEs (red). Three of four replicates show no significant difference in variability of poly-A signal across bins within features (Wilcoxon rank-sum test P>0.05). Replicate 3 shows a significant difference (Wilcoxon rank-sum test, P=0.047). **E)** Bar plot shows number of genic fusions reproducibly found in poly-A RNA-seq data. **F)** For each TE introduced in **E**, the y-axis position of each point represents the number of uniquely-mapping chimeric reads detected by STAR that support each breakpoint (see **Methods**). 

\newpage

```{r, fig.dim=c(8,10), out.height="6.4in", out.width="8in"}
read_rds("~/work/TestisTEs2021/results/panels/supp3.rds")
```

*Supplementary Figure 3. Gene Expression Program detection via consensus approach to Independent Component Analysis.* 
**A)** Heatmaps correspond to independent replicates of the grid search approach used to optimize ICA component number (*k*) and *q*-value cutoff. Color intensity corresponds to the enrichment score (see **Methods**) for each combination of *q* and *k*. **B)** Histogram shows module sizes among the set of GEPs used in main analysis. The majority of modules detected include fewer than 100 features. **C)** Barplots show percentage of discovered GEPs with at least 1 unique Biological Process, Cellular Component, or Molecular Function enrichment. **D)** Scatterplots show relationship of TE expression in *w1118* Cluster 3/Spermatocyte versus all clusters detected by our pipeline in the "Wild Strain" dataset generated by Witt et al. 2019. Multiple spermatocyte clusters (highlighted in red) show TE expression patterns similar to 3/Spermatocyte.

\newpage

```{r, fig.dim=c(8,8), out.height="7in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/supp4.rds")
```

*Supplementary Figure 4. Redundancy and TE content of GEPs.* **A)** Clustering of GEPs by cell usage score (consensus ICA source matrix). **B)** Clustering of GEPs by gene membership score (consensus ICA mixing matrix). For **A**,**B** `1-Pearson's R` is used as a distance metric. **C)** Breakdown of specific DNA TE, LINE, and LTR TE families in GEP-27.


\newpage

```{r, fig.dim=c(3.5,4), out.height="4in", out.width="7in", fig.align="center", dpi=72, dev="png"}
image_read('/media/mlawlor/T7/microscopy_figs/210215_accord2_calfluor610_eachm_quasar670_3p4.slices_1_1.split_channels.png',strip = T, density = 200) %>% image_ggplot()
```

*Supplementary Figure 5. _ACCORD2_ and _EAChm_ expression in L3 testes.*
Representative slice of multiplexed RNA-FISH in whole-mount 3rd larval instar w1118 testis. Image is split by color channel. *ACCORD2* and *EAChm* expression is detected in the middle region of the testis, where primary spermatocytes are located. Red: *EAChm*; green: *ACCORD2*; blue: DAPI. 

\newpage

```{r, fig.dim=c(3.5,4), out.height="5in", out.width="7in", fig.align='center', dev="png"}
image_read('/media/mlawlor/T7/microscopy_figs/210215_accord2_calfluor610_eachm_quasar670_3p4.slices_1_30.montage.slices_1_30.montage.png',strip = T,density = 200) %>% image_ggplot()
```

*Supplementary Figure 6. _ACCORD2_ and _EAChm_ expression in L3 testes.* 
Z-slices montage of multiplexed RNA-FISH in whole-mount 3rd larval instar w1118 testis. *ACCORD2* and *EAChm* expression is detected in the middle region of the testis, where primary spermatocytes are located. Red: *EAChm*; green: *ACCORD2*; blue: DAPI. 

\newpage

```{r, fig.dim=c(3.5,4), out.height="5in", out.width="7in", fig.align='center', dev="png"}
s7a <- image_read('/media/mlawlor/T7/microscopy_figs/210204_dapi_quasimodo2_calfluor61_eachm_quasar670p.slices_1_1.representative.png',strip = T,density = 200) %>% image_ggplot()

#s7b <- image_read('/media/mlawlor/T7/microscopy_figs/210204_dapi_quasimodo2_calfluor61_eachm_quasar670p.slices_1_22_x2.montage.png',strip = T,density = 200) %>% image_ggplot()

s7a

```

*Supplementary Figure 7. _QUASIMODO2_ expression in L3 testes.* 
Representative RNA-FISH z-slice in whole-mount 3rd larval instar *w1118* testis. *QUASIMODO2* expression is detected in the middle region of the testis, where primary spermatocytes are located. Green: *QUASIMODO2*; blue: DAPI. 

\newpage

```{r, fig.dim=c(10,10), out.height="7in", out.width="7in"}
read_rds("~/work/TestisTEs2021/results/panels/supp5.rds")
```

*Supplementary Figure 8. TEP-TEs are enriched on the Y-chromosome.* 
**A)** Dot plot shows expression of selected effectors of spermatocyte transcriptional programs. Color of each dot corresponds to mean normalized and log-transformed expression within cell clusters. Dot size corresponds to the proportion of cells in each cluster expressing the marker. **B)** Bars show strength of top 10 enriched Gene Ontology Molecular Function terms for GEP-27. **C)** Bar plot shows proportion of Y chromosome genes or other genes that are assigned to the TE-enriched Program or other GEPs. Chi-square test P=1.7e-05. **D)** Barplot shows percentage of TEP or non-TEP TEs with at least 1 Y-linked insertion detected by RepeatMasker in the heterochromatin-enriched assembly described by Chang and Larracuente 2019. **E)** Bars represent the ratios of male to female estimated copies for the top 10 male- and female-enriched TEs. Bars are colored by membership in TEP. **F)** Allele specific analysis of TE expression (see **Methods**) shows that non-Y-linked copies of TEP TEs are expressed proportionately to their DNA copy number. Wilcoxon rank-sum test, P=0.24.
