---
title: "Figures"
author: ""
date: "4/27/2021"
output:
  pdf_document:
    latex_engine: xelatex
sansfont: Arial
mainfont: Arial
header-includes:
- \AtBeginDocument{\let\maketitle\relax}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, warning = F, message = F, dev = "png", dpi = 200)
```

```{r, include=F}
library(tidyverse)
library(patchwork)
library(ggtext)
library(magick)

extrafont::loadfonts(quiet=TRUE)
```

\newpage


## Supplementary Figure 1

```{r, fig.dim=c(20,20), out.height="6.5in", out.width="6.5in"}
read_rds("results/panels/supp1.rds")
```

***Supplementary Figure 1. scRNA-seq preprocessing and comparison to prior analysis.***  **A)** Barplot shows the number of cells post-filtering in Mahadevaraju et al. 2021 (gray) and this study (red). **B)** Heatmap shows correlation between scRNA-seq expression estimates derived from our pipeline for all clusters identified in this study compared to clusters identified by Mahadevaraju et al. 2021 (source of data).  **C)** Dotplot shows scaled expression of selected spermatocyte and spermatogonia markers.  

\newpage

## Supplementary Figure 2

```{r, fig.dim=c(20,20), out.height="7in", out.width="7in",fig.align='center'}
read_rds("results/panels/supp2.rds")
```

***Supplementary Figure 2. TE expression in scRNA-seq data from larval testes.*** **A)** Violin plots show distributions of raw TE-mapping UMI counts in each w1118 L3 testis cell cluster. TE counts vary significantly among the clusters (Kruskal-Wallis, p < 2.2e-16). **B)** Violin plots show distributions of depth normalized TE-mapping UMI counts in each w1118 L3 testis cell cluster. TE counts vary significantly among the clusters (Kruskal-Wallis, p < 2.2e-16). **C)** Scatterplots show pseudo-bulk expression derived from our w1118 scRNA pipeline (see Methods) versus bulk expression for four w1118 testis poly-A RNA-seq replicates generated by Mahadevaraju et al. 2021 (see Methods). Each replicate shows strong correlation with pseudo-bulk (all Pearson's R >= 0.89, P<2.2e-16). **D)** Scatterplots show same analysis described in **D** but restricted to TEs (all Pearson's R >= 0.85, P<2.2e-16). **E)** Heatmaps show distribution of sense-strand bulk poly-A RNA-seq signal for single-isoform host gene mRNAs (left) and detected TEs (right). Larval bulk poly-A RNA-seq data generated by Mahadevaraju et al. 2021 was mapped to each feature (see Methods) and divided into bins representing one tenth the full length of each feature. **F)** Boxplots show standard deviations of expression across bins for host genes (gray) and TEs (red). Three of four replicates show no significant difference in variability of poly-A signal across bins within features (Wilcoxon rank-sum test P>0.05). Replicate 3 shows a significant difference (Wilcoxon rank-sum test, two-sided P=0.047). For all boxplots, midline represents median, box represents interquartile range (IQR), and whiskers extend >1.5 IQR from the upper or lower quartile. **G)** Bar plot shows number of TE-gene chimeric transcripts reproducibly found in bulk poly-A RNA-seq data, for all TEs with a least one chimeric transcript identified in at least two of the replicate RNA-seq datasets. **H)** For each TE introduced in G, the y-axis position of each point represents the number of uniquely-mapping chimeric reads detected by STAR that support each chimeric transcript (see Methods). 

\newpage

## Supplementary Figure 3

```{r, fig.dim=c(6,8), out.width="6in", out.height="6in",fig.align='center'}
read_rds("results/panels/supp3.rds")
```

***Supplementary Figure 3. Co-expression module detection via consensus approach to Independent Component Analysis (cICA).*** **A)** Final module gene membership scores were compared to modules identified in replicate cICA runs by Spearman correlation. For each module, the correlation p-values for the best correlated module from each cICA replicate are plotted. Modules (x axis) are ordered by the Spearman correlation with the best match across all three cICA replicates. **B)** Histogram shows distribution of the number of genes or TEs assigned to the set of modules used in main analysis.  **C)** Barplots show percentage of modules uniquely enriched for at least one Biological Process (BP), Cellular Component (CC), or Molecular Function (MF) Gene Ontology Term.  

\newpage


## Supplementary Figure 4

```{r, fig.dim=c(3.5,4), out.height="4in", out.width="7in", fig.align="center", dpi=72, dev="png"}
image_read('resources/210215_accord2_calfluor610_eachm_quasar670_3p4.slices_1_1.split_channels.png',strip = T, density = 200) %>% image_ggplot()
```

***Supplementary Figure 4. ACCORD2 and EAChm expression in L3 testes.*** Representative slice of multiplexed RNA-FISH in whole-mount 3rd larval instar w1118 testis. Image is split by color channel. *ACCORD2* and *EAChm* expression is detected in the middle region of the testis, where primary spermatocytes are located. Red: *EAChm*; green: *ACCORD2*; blue: DAPI. 

\newpage

## Supplementary Figure 5

```{r, fig.dim=c(3.5,4), out.height="5in", out.width="7in", fig.align='center', dev="png"}
image_read('resources/210215_accord2_calfluor610_eachm_quasar670_3p4.slices_1_30.montage.slices_1_30.montage.png',strip = T,density = 200) %>% image_ggplot()
```

***Supplementary Figure 5. ACCORD2 and EAChm expression in L3 testes.*** Z-slices montage of multiplexed RNA-FISH in whole-mount 3rd larval instar w1118 testis. *ACCORD2* and *EAChm* expression is detected in the middle region of the testis, where primary spermatocytes are located. Red: *EAChm*; green: *ACCORD2*; blue: DAPI. 

\newpage

## Supplementary Figure 6

```{r, fig.dim=c(3.5,4), out.height="5in", out.width="7in", fig.align='center', dev="png"}
s7a <- image_read('resources/210204_dapi_quasimodo2_calfluor61_eachm_quasar670p.slices_1_1.representative.png',strip = T,density = 200) %>% image_ggplot()

#s7b <- image_read('resources/210204_dapi_quasimodo2_calfluor61_eachm_quasar670p.slices_1_22_x2.montage.png',strip = T,density = 200) %>% image_ggplot()

s7a

```

***Supplementary Figure 6. QUASIMODO2 expression in L3 testes.*** Representative RNA-FISH z-slice in whole-mount 3rd larval instar w1118 testis. *QUASIMODO2* expression is detected in the middle region of the testis, where primary spermatocytes are located. Green: *QUASIMODO2*; blue: DAPI. 

\newpage

## Supplementary Figure 7

```{r, fig.dim=c(8,4), out.height="5in", out.width="5in",fig.align='center'}
read_rds("results/figs/all_dataset_tep_scores/all_dataset_tep_scores.tes.ggp.rds") + 
  theme(aspect.ratio = NULL, axis.title.x = element_markdown(size=rel(0.5)),strip.text = element_text(size=rel(0.7))) +
  theme(plot.title = element_text(size=rel(0.5)))
```

***Supplementary Figure 7. TE Expression in adult data.*** Scatterplots show relationship of TE expression in w1118 Cluster 3/Spermatocytes versus all clusters detected by our pipeline in the "Wild Strain" dataset generated by Witt et al. 2019. Multiple spermatocyte clusters (highlighted in red) show TE expression patterns similar to 3/Spermatocytes. 

\newpage

## Supplementary Figure 8

```{r, fig.dim=c(10,12), out.height="7in", out.width="5.81in"}
read_rds("results/panels/supp5.rds")
```

***Supplementary Figure 8. Module 27 TEs are enriched on the Y-chromosome.*** **A)** Dot plot shows expression of selected effectors of spermatocyte transcriptional programs. Color of each dot corresponds to mean normalized and log-transformed expression within cell clusters. Dot size corresponds to the proportion of cells in each cluster expressing the marker. **B)** Bars show strength of top 10 enriched Gene Ontology Molecular Function terms for module 27. **C)** Bar plot shows proportion of Y chromosome genes or other genes used for module detection analysis that are assigned to module 27 or other modules. Fisher’s Exact test, two-sided P=1.7e-05. **D)** Barplot shows percentage of TEs assigned to module 27 or other TEs with at least one Y-linked insertion detected by RepeatMasker in the heterochromatin-enriched assembly described by Chang and Larracuente 2019. **E)** Allele specific analysis of TE expression (see Methods) shows that non-Y-linked copies of module 27 TEs are expressed proportionately to their DNA copy number. Wilcoxon rank-sum test, two-sided P=0.24.  

\newpage


## Supplementary Figure 9


```{r, fig.dim=c(8,4), out.height="5in", out.width="5in",fig.align='center'}
read_rds( "results/panels/supp4_5.rds")
```

***Supplementary Figure 9. Silencing of module 27 TEs in non-spermatocyte contexts.*** **A)** Barplot shows the percentage of module 27 or other TEs with fragments in the *flamenco* locus. Module 27 is significantly enriched for TEs represented in the *flamenco* locus (Fisher’s Exact Test, two-sided p = 0.03). **B)** Barplot shows the percentage of module 27 or other TEs which are silenced by the piRNA pathway in *Drosophila* ovaries. Module 27 is significantly enriched for TEs silenced in ovaries (Fisher’s Exact Test, two-sided P=7.6e-4). 

\newpage

## Supplementary Figure 10

```{r, fig.dim=c(10,7), out.height="3in", out.width="7in", align="center", fig.align='center'}
read_rds("results/figs/arriba_total_rna_fusions/arriba_total_rna_fusions.ggp.rds")
```

***Supplementary Figure 10. Detected fusions between module 27 TEs and Y-linked genes.*** Fusions between module 27 TEs and Y-linked genes detected by the Arriba fusion caller in 2 or more w1118 adult testis total RNA-seq replicates. Y-axis shows the percentage of reads that align to the breakpoint that support the fusion call (split reads and discordant pairs). 

\newpage

## Supplementary Figure 11

```{r, fig.dim=c(10,7), out.height="3in", out.width="7in", align="center", fig.align='center'}
read_rds("results/figs/tidal_larval_tep_xa_boxplot/tidal_larval_tep_xa_boxplot.4.ggp.rds")
```

***Supplementary Figure 11.*** Chr4-linked versus autosomal polymorphic insertions. Boxplot showing the ratios of chr4-linked versus autosomal polymorphic insertions per mappable megabase for each TE in the TIDAL-fly database. Module 27 TEs show similar levels of depletion on the 4th chromosome compared to other TEs with polymorphic insertions (Wilcoxon rank-sum test, P = 0.34). For all boxplots, midline represents median, box represents interquartile range (IQR), and whiskers extend >1.5 IQR from the upper or lower quartile. 

\newpage

## Supplementary Figure 12

```{r, fig.dim=c(10,7), out.height="3in", out.width="7in", align="center", fig.align='center'}
read_rds("results/figs/larval_scrna_basic_qc_stats/larval_scrna_basic_qc_stats.batch.ggp.rds") + theme(aspect.ratio = NULL)+
  theme(axis.title = element_text(size=rel(1)))
```

***Supplementary Figure 12. Cells in scRNA-seq clusters.*** Barplot shows the number of cells assigned to each cluster. Bars are divided and colored by the scRNA replicate from which cells are derived. 

\newpage

## Supplementary Figure 13

```{r, fig.dim=c(8,8), out.height="7in", out.width="7in"}
read_rds("results/panels/supp4.rds")
```

***Supplementary Figure 13. Redundancy and TE content of modules.*** **A)** Clustering of modules by cell usage score (consensus ICA source matrix). **B)** Clustering of modules by gene membership score (consensus ICA mixing matrix). For A and B `1-Pearson's R` is used as a distance metric. 


